### ISSUE TITLE
STYLE-WARNING: redefining EMACS-INSPECT (#<SB-PCL:SYSTEM-CLASS T>) in DEFMETHOD

### ISSUE THREAD

**Issue by Nact** [slime/slime#230]
I get this message when starting SLIME:

; loading #P"d:/lisp/slime-2.12/swank-loader.lisp"
STYLE-WARNING: redefining EMACS-INSPECT (#<SB-PCL:SYSTEM-CLASS T>) in DEFMETHOD

The REPL seems to work normal.

I am totally new to EMACS and SLIME and I'd like to know what does this mean and how to fix it.
I use Windows 8 (64-bit), GNU Emacs 25.0.50.1 (x86_64-w64-mingw32), SLIME 2.12, and SBCL 1.2.7

**Comment by jackcarrozzo** [comment]
That's ok- it's just letting you know that a previously-defined symbol is being redefined.

**Comment by Nact** [comment]
Thank you jackcarrozzo! Could you describe briefly what EMACS-INSPECT is responsible for? And is there a good source to learn slime compiler messages/errors/interface?

**Comment by jackcarrozzo** [comment]
Good question - it took a bit of spelunking to find a full answer. The short answer is that emacs-inspect stores handlers for inspecting objects; when a particular object is inspected, the relevant generic function (based on type) will match and be evaluated. From http://lisp-book.org/contents/chslime.pdf page 16, you can run this to see the currently-attached generics:

````M-.  swank-backend:emacs-inspect````

That pdf also describes creating your own inspector as well as additional features that look pretty cool.

Regarding your second point: lots of stuff gets printed to the *slime-events* buffer; unless you're having a slime-specific issue, you probably don't need to even have it open in a window. Warnings and so forth relevant to your code and interactions will come out either directly in the REPL, in the *inferior-lisp* buffer, or in one of a few other buffers that emacs/slime will auto-open for you when needed.

````
CL-USER> (use-package :elk)
; Evaluation aborted on #<SB-KERNEL:SIMPLE-PACKAGE-ERROR "The name ~S does not designate any package." {1002C9D683}>.
CL-USER> (defun moose (a) (+ a 7))
MOOSE
CL-USER> (defun moose (a) (+ a 8))
STYLE-WARNING: redefining COMMON-LISP-USER::MOOSE in DEFUN
MOOSE
CL-USER>
````

So, in summary: don't worry about it. Slime makes interfacing Common Lisp in emacs simple, and it does a great job of staying out of the way. Slime does admittedly have a ton of features but it doesn't force you to use them. Note to self: I should really get around to learning them some day...

**Comment by Nact** [comment]
Jack, thank you so much! It is a really big deal for me since I am a beginner in emacs/slime world yet I am a serious guy looking for a professional approach in this field. I am going to repost your valuable advice at stackoverflow if you don't mind.
Good luck with your Z80!

**Comment by jackcarrozzo** [comment]
No problem - feel free to repost whereever. If you're just getting started with lisp, I found these two books to be most useful:

- Peter Seibel: Practical Common Lisp
- Peter Norvig: Paradigms of Artificial Intelligence Programming: Case Studies in Common Lisp

To supplement learning the basics with modern webapp projects, check out these packages by @fukamachi:

- [http://8arrow.org/ningle/](http://8arrow.org/ningle/)
- [http://8arrow.org/caveman/](http://8arrow.org/caveman/)

I've used them both - they are surprisingly quick to use, and the docs are good.

I've been slowly writing a tiny realtime lisp OS for the Z80 machines, maybe I will actually get it to a usable point someday ;-)

**Comment by Nact** [comment]
Practical Common Lisp is a very nice book really, I am going to take the web direction of lisp developing in future, so lucky you posted these links, and now I have a good point to start.
I could not start experimenting until you had explained me about the compiler message which worried me much.
Thank you from the bottom of my heart.
I reposted your reply to stackoverflow, I guess many people seek the answer too.

**Comment by luismbo** [comment]
Here's one way to fix that warning.

```diff
--- a/swank/backend.lisp
+++ b/swank/backend.lisp
@@ -1232,15 +1232,16 @@ inserted into the buffer as is, or a list of the form:
  after calling the lambda.
 "))
 
-(defmethod emacs-inspect ((object t))
-  "Generic method for inspecting any kind of object.
+(defmethod no-applicable-method ((fn (eql #'emacs-inspect)) &rest args)
+  "Fallback for inspecting any kind of object.
 
 Since we don't know how to deal with OBJECT we simply dump the
 output of CL:DESCRIBE."
-   `("Type: " (:value ,(type-of object)) (:newline)
-     "Don't know how to inspect the object, dumping output of CL:DESCRIBE:"
-     (:newline) (:newline)
-     ,(with-output-to-string (desc) (describe object desc))))
+  (let ((object (first args)))
+    `("Type: " (:value ,(type-of object)) (:newline)
+      "Don't know how to inspect the object, dumping output of CL:DESCRIBE:"
+      (:newline) (:newline)
+      ,(with-output-to-string (desc) (describe object desc)))))
 
 (definterface eval-context (object)
   "Return a list of bindings corresponding to OBJECT's slots."
```

Not 100% sure it's a good idea to specialize `no-applicable-method` like this. @ellerh?

**Comment by Nact** [comment]
That is food for thought. As far as I can understand there is a method or object after name '#' which belongs to 'emacs-inspect' object and is being modified by slime compiler on slime start. I can not by now be sure which of '#' out of these had been redefined by compiler:
Backtrace:
  0: (SB-INT:SIMPLE-EVAL-IN-LEXENV SWANK/BACKEND:EMACS-INSPECT #<NULL-LEXENV>)
  1: (EVAL SWANK/BACKEND:EMACS-INSPECT)
  2: (INTERACTIVE-EVAL SWANK/BACKEND:EMACS-INSPECT :EVAL NIL)
  3: (SB-IMPL::REPL-FUN NIL)
  4: ((LAMBDA NIL :IN SB-IMPL::TOPLEVEL-REPL))
  5: (SB-IMPL::%WITH-REBOUND-IO-SYNTAX #<CLOSURE (LAMBDA NIL :IN SB-IMPL::TOPLEVEL-REPL) {100310B21B}>)
  6: (SB-IMPL::TOPLEVEL-REPL NIL)
  7: (SB-IMPL::TOPLEVEL-INIT)
  8: ((FLET #:WITHOUT-INTERRUPTS-BODY-87 :IN SAVE-LISP-AND-DIE))
  9: ((LABELS SB-IMPL::RESTART-LISP :IN SAVE-LISP-AND-DIE))
 10: ("foreign function: #x42E2AC")

**Comment by luismbo** [comment]
Hmm. A lone # usually means some object was elided while being printed. Can you paste the actual error message that came with that backtrace?

**Comment by Nact** [comment]
SBCL is free software, provided as is, with absolutely no warranty.
It is mostly in the public domain; some portions are provided under
BSD-style licenses.  See the CREDITS and COPYING files in the
distribution for more information.

WARNING: the Windows port is fragile, particularly for multithreaded
code.  Unfortunately, the development team currently lacks the time
and resources this platform demands.
* 
; loading #P"d:/lisp/slime-2.12/swank-loader.lisp"
STYLE-WARNING: redefining EMACS-INSPECT (#<SB-PCL:SYSTEM-CLASS T>) in DEFMETHOD
I just noticed that the piece of code in my first post is incorrect '(#<SB-PCL:SYSTEM-CLASS T>)' is the right output instead of just '(#)'. Though the name of the topic is with correct output. Still I do not see SB-PCL:SYSTEM-CLASS in the backtrace nor I can guess what it is.

**Comment by ellerh** [comment]
In the past, `no-applicable-method` was buggy in Allegro CL.  Maybe it's fixed now.

An alternative idea would be to intdroduce a new backend function, say `inspector-for-object`
that can wrap the to be inspected object in some other class (like the CCL inspector does).  By default `inspector-for-object` would just return the object but for SBCL it could could return some
other class. Instead of calling `emacs-inspect` directly we would call `(emacs-inspect (inspector-for-object X))`.  Not sure if the added complexitiy is worth it.

**Comment by earl-ducaine** [comment]
Just adding my own opinion: obviously a low priority, as it's just a platform style warning. 

But it is in fact bad style as it can introduce subtle bugs, e.g. during refactoring the load ordering of the two method definitions is reversed. 

Also, it clutters the output of loading and compiling with warnings, making it more difficult for users of the library to find the warning they really do care about.
